src/rigid_pendulum_poc/solver.h

```diff
@@ -4,14 +4,138 @@
+
+/// @brief A class to store and manage the states of a dynamic system
+// TODO: Refactor this class to avoid/minimize expensive copies
+class State {
+public:
+    State();
+    State(HostView1D, HostView1D, HostView1D, HostView1D);
+
+    /// Get the generalized coordinates
+    inline HostView1D GetGeneralizedCoordinates() const { return generalized_coords_; }
+
+    /// Get the first time derivative of the generalized coordinates
+    inline HostView1D GetGeneralizedVelocity() const { return generalized_velocity_; }
+
+    /// Get the second time derivative of the generalized coordinates
+    inline HostView1D GetGeneralizedAcceleration() const { return generalized_acceleration_; }
+
+    /// Get the algorithmic accelerations (different than the generalized accelerations)
+    inline HostView1D GetAlgorithmicAcceleration() const { return algorithmic_acceleration_; }
+
+private:
+    HostView1D generalized_coords_;
+    HostView1D generalized_velocity_;
+    HostView1D generalized_acceleration_;
+    HostView1D algorithmic_acceleration_;
+};
+
+HostView2D create_identity_matrix(size_t size);
+HostView1D create_identity_vector(size_t size);
+
+/// @brief A time integrator class based on the generalized-alpha method
+class GeneralizedAlphaTimeIntegrator {
+public:
+    static constexpr double kTOLERANCE = 1e-6;
+
+    GeneralizedAlphaTimeIntegrator(
+        double initial_time = 0., double time_step = 1., size_t n_steps = 1, double alpha_f = 0.5,
+        double alpha_m = 0.5, double beta = 0.25, double gamma = 0.5, size_t max_iterations = 10
+    );
+
+    /// Returns the initial time of the analysis
+    inline double GetInitialTime() const { return initial_time_; }
+
+    /// Returns the current time of the analysis
+    inline double GetCurrentTime() const { return current_time_; }
+
+    /// Returns the time step of the analysis
+    inline double GetTimeStep() const { return time_step_; }
+
+    /// Advances the current analysis time by one time step
+    inline void AdvanceTimeStep() { current_time_ += time_step_; }
+
+    /// Returns the number of analysis time steps
+    inline int GetNumberOfSteps() const { return n_steps_; }
+
+    /// Returns the alpha_f parameter
+    inline double GetAlphaF() const { return kALPHA_F_; }
+
+    /// Returns the alpha_m parameter
+    inline double GetAlphaM() const { return kALPHA_M_; }
+
+    /// Returns the beta parameter
+    inline double GetBeta() const { return kBETA_; }
+
+    /// Returns the gamma parameter
+    inline double GetGamma() const { return kGAMMA_; }
+
+    /// Returns the maximum number of iterations
+    inline size_t GetMaxIterations() const { return kMAX_ITERATIONS_; }
+
+    /// Performs the time integration and returns a vector of States over the time steps
+    std::vector<State> Integrate(
+        const State&, std::function<HostView2D(size_t)> iteration_matrix = create_identity_matrix,
+        std::function<HostView1D(size_t)> residual_vector = create_identity_vector
```

Comment:
This should work to let us customize these calculations.  

Usually, these function objects would be passed in at construction time and stored, usually as a single class that defines two different methods.  However, let's continue with what you have now and see how it looks in the end.  

We might also need to revisit if the `std::function<...>` should be passed by reference.  Type erasure (the technique used to implement it) is complex, so reference vs value may have both performance and correctness implications.

Reply:
Sounds good to me.